src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs

```diff
@@ -129,29 +129,57 @@ public void OnResponseHeader(ReadOnlySpan<byte> name, ReadOnlySpan<byte> value)
                         throw new Http2ProtocolException(Http2ProtocolErrorCode.ProtocolError);
                     }
 
-                    if (name.SequenceEqual(s_statusHeaderName))
+                    if (name[0] == 58) // ':'
                     {
                         if (_state == StreamState.ExpectingTrailingHeaders)
                         {
                             // Pseudo-headers not allowed in trailers.
                             if (NetEventSource.IsEnabled) _connection.Trace("Pseudo-header in trailer headers.");
                             throw new HttpRequestException(SR.net_http_invalid_response);
                         }
-
-                        if (value.Length != 3)
-                            throw new Exception("Invalid status code");
-
-                        // Copied from HttpConnection
-                        byte status1 = value[0], status2 = value[1], status3 = value[2];
-                        if (!IsDigit(status1) || !IsDigit(status2) || !IsDigit(status3))
+                        if (name.SequenceEqual(s_statusHeaderName))
+                        {
+                            if (_response.StatusCode != 0)
+                            {
+                                if (NetEventSource.IsEnabled) _connection.Trace("Received duplicate status headers.");
+                                throw new HttpRequestException(SR.net_http_invalid_response);
+                            }
+
+                            if (value.Length != 3)
+                            {
+                                if (NetEventSource.IsEnabled) _connection.Trace($"Received status headers '{System.Text.Encoding.ASCII.GetString(name)}'.");
+                                throw new HttpRequestException(SR.net_http_invalid_response);
+                            }
+
+                            // Copied from HttpConnection
+                            byte status1 = value[0], status2 = value[1], status3 = value[2];
+                            if (!IsDigit(status1) || !IsDigit(status2) || !IsDigit(status3))
+                            {
+                                if (NetEventSource.IsEnabled) _connection.Trace($"Received status headers '{System.Text.Encoding.ASCII.GetString(name)}'.");
+                                throw new HttpRequestException(SR.net_http_invalid_response);
+                            }
+
+                            int statusValue = (100 * (status1 - '0') + 10 * (status2 - '0') + (status3 - '0'));
+                            if (statusValue == 0)
+                            {
+                                throw new Exception("Invalid status code zero.");
```

Comment:
We shouldn't be throwing 'Exception' type?  This should be something like HttpRequestException.

Reply:
And why are these literal strings and not from the SR.* string resources?

Reply:
Good point. There was existing exception I copy but I think that was bad example.